core/src/main/scala/iliad/gl/opengl.scala

```diff
@@ -0,0 +1,223 @@
+package iliad
+package gl
+
+import iliad.kernel.platform.GLES30Library
+import iliad.CatsExtra._
+import iliad.std.int._
+
+import cats._
+import cats.data._
+import cats.free._, Free._
+import cats.implicits._
+
+import java.nio.Buffer
+
+object GL {
+
+  type DSL[A] = Free[GL, A]
+  type Effect[F[_], A] = ReaderT[F, GLES30Library, A]
+
+  type NoEffect[A] = Effect[Id, A]
+  type LogEffect[A] = Effect[Writer[List[String], ?], A]
+  type DebugEffect[A] = Effect[XorT[Writer[List[String], ?], String, ?], A]
+
+  type Interpreter[F[_]] = GL ~> F
+
+  val run: Interpreter[NoEffect] = GLInterpreter
+  val log: Interpreter[LogEffect] = GLLogInterpreter
+  val debug: Interpreter[DebugEffect] = GLDebugInterpreter
+
+  def interpret[F[_]: Monad](f: Interpreter[F]): (DSL ~> F) = new (DSL ~> F) {
+    def apply[A](gl: DSL[A]): F[A] = gl.foldMap(f)
+  }
+
+  val getError: DSL[Int] = GLGetError.free
+
+  private def getShaderLogLength(shader: Int): DSL[Int] =
+    GLGetShaderiv(shader, GL_INFO_LOG_LENGTH).free
+  private def getCompileStatus(shader: Int): DSL[Int] =
+    GLGetShaderiv(shader, GL_COMPILE_STATUS).free
+
+  def getCompileError(shader: Int): DSL[Option[String]] =
+    getCompileStatus(shader) flatMap { status =>
+      if (status == GL_TRUE.value) Free.pure(None)
+      else
+        for {
+          l <- getShaderLogLength(shader)
+          s <- liftF(GLGetShaderInfoLog(shader, l))
+        } yield Some(s)
+    }
+
+  def makeVertexShader(source: String): DSL[Int] =
+    for {
+      id <- GLCreateShader(GL_VERTEX_SHADER).free
+      _ <- GLShaderSource(id, List(source)).free
+      _ <- GLCompileShader(id).free
+    } yield id
+
+  def makeFragmentShader(source: String): DSL[Int] =
+    for {
+      id <- GLCreateShader(GL_FRAGMENT_SHADER).free
+      _ <- GLShaderSource(id, List(source)).free
+      _ <- GLCompileShader(id).free
+    } yield id
+
+  def makeProgram(vertexId: Int, fragmentId: Int): DSL[Int] =
+    for {
+      id <- GLCreateProgram.free
+      _ <- GLAttachShader(id, vertexId).free
+      _ <- GLAttachShader(id, fragmentId).free
+      _ <- GLLinkProgram(id).free
+    } yield id
+
+  private def traverse[A, B, G[_]: Traverse](keys: G[A])(
+      f: A => DSL[B]): DSL[G[(A, B)]] =
```

Comment:
Please call `traverseKeys` or a different name since the signature is not the same as traverse.